### PR TITLE
Add JS remote logging package

### LINKED ISSUE #49284: [Remote Error Logging] Create JS remote error logging package
## Description

Develop a JavaScript package for remote error logging to be reused in WooCommerce Admin and WooCommerce Block. This package should capture and log JS errors and send them to a centralized logging system. The package should:

- Capture `window.onerror` and `unhandledrejection` calls and send them to the API automatically if the feature is enabled.
- Provide a log function to log unusual events for hard-to-debug cases manually.
- Filter out errors originating from third parties by dropping error events that exclusively contain third-party stack frames.

Refer to Calypso's [catch-js-errors](https://github.com/Automattic/wp-calypso/tree/trunk/client/lib/catch-js-errors) module for implementation details

## Accept Criteria

1. A JavaScript package is created for remote error logging.
2. The package captures `window.onerror` and `unhandledrejection` calls and sends them to the API automatically if logging is enabled.
3. The package exports a log function for manually logging unusual events.
4. The package accepts a regular pattern or function as a parameter to filter out errors originating from third parties by dropping error events that exclusively contain third-party stack frames.
5. The package is documented, including usage instructions.
6. The package is unit-tested to ensure it works as expected.

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #49284.

This PR introduces a new JavaScript package for remote error logging in WooCommerce. The package captures and logs JS errors, sending them to our centralized logging system. 

Key features include:

- Automatic capture of `window.onerror` and `unhandledrejection` events
- A manual log function for unusual events
- Error filtering to exclude third-party stack frames
- Rate limiting to prevent API flooding
- Customizable through WordPress filters

See [REDAME.md ](https://github.com/woocommerce/woocommerce/blob/add/js-remote-error-logging/packages/js/remote-logging/REDAME.md) for more details.

This PR only adds a new package, it does not change any existing functionality. We will have a separate PR to integrate the new package with WC Admin or WC Blocks to capture and log errors from the WC JavaScript code.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**For QA team, this PR only adds a new package, it does not change any existing functionality so you can skip doing manual testing for this PR.** 
 
1. Sandbox public-api.wordpress.com and use D156767-code patch if it has not been merged yet.
2. Checkout out [https://github.com/woocommerce/woocommerce/tree/test/wcadmin-remote-js-logging](wcadmin-remote-js-logging) branch or cherry-pick https://github.com/woocommerce/woocommerce/commit/093ba079fd84d90fd4c301a39db11e86c225a17e
3. Run `pnpm install`
4. Run `pnpm watch:build` on the `./plugins/woocommerce` folder.
5. Use local fresh site or delete `woocommerce_allow_tracking` option to test the changes.
6. Open browser dev console and run `window.localStorage.setItem( 'debug', 'wc:remote-logging')` to enable the debug logging.
7. Go to `WooCommerce > Home`
5. Confirm `wc:remote-logging Tracks is not enabled.` message is logged in the console.
6. Enable Tracks by adding `wp option set woocommerce_allow_tracking yes`
8. Reload the page
9. Confirm `wc:remote-logging Caught error event` and `wc:remote-logging Sending error to API` messages are logged in the console.
10. Check the network tab in dev console and confirm a `POST` request is made to the remote logging API.
11. Reload the page
12. Confirm `Rate limit reached. Skipping send error` message is logged in the console.
13. Comment out `setTimeout` in [`plugins/woocommerce-admin/client/lib/init-remote-logging.ts`](https://github.com/woocommerce/woocommerce/blob/093ba079fd84d90fd4c301a39db11e86c225a17e/plugins/woocommerce-admin/client/lib/init-remote-logging.ts) and uncomment `new Promise to test the asynchronous error handling.
14. Reload the page
15. Confirm `wc:remote-logging Caught unhandled rejection:` message is logged in the console.
16. Check the network tab in the dev console and confirm a `POST` request is made to the remote logging API.
17. Comment out `new Promis` in `plugins/woocommerce-admin/client/lib/init-remote-logging.ts`
18. Uncomment `log` in `plugins/woocommerce-admin/client/lib/init-remote-logging.ts`
19. Reload the page
20. Confirm `wc:remote-logging Sending log to API` message is logged in the console.
21. Check the network tab in dev console and confirm a `POST` request is made to the remote logging API.
22.  Please check `D156767` patch for how to view the logs in Kibana.

Error log:
Screenshot 2024-07-29 at 17 54 02

Info log:
Screenshot 2024-07-29 at 17 54 59

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Add JS remote logging package

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: packages/js/remote-logging/src/remote-logger.ts [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+	 */
+	private async sendLog( logData: LogData ): Promise< void > {
+		const body = new window.FormData();
+		body.append( 'params', JSON.stringify( logData ) );
+
+		try {
+			debug( 'Sending log to API:', logData );
+
+			await window.fetch(
+				'https://public-api.wordpress.com/rest/v1.1/logstash',
```
Comment (rjchow) [inline packages/js/remote-logging/src/remote-logger.ts]: should we add a filter to customise this endpoint?

File: packages/js/remote-logging/src/remote-logger.ts [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+	 * @param error - The error data to be sent.
+	 */
+	private async sendError( error: ErrorData ) {
+		const body = new window.FormData();
+		body.append( 'error', JSON.stringify( error ) );
+
+		try {
+			debug( 'Sending error to API:', error );
+			await window.fetch(
+				'https://public-api.wordpress.com/rest/v1.1/js-error',
```
Comment (rjchow) [inline packages/js/remote-logging/src/remote-logger.ts]: should we add a filter to customise this endpoint?

File: packages/js/remote-logging/src/test/utilts.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
+/**
```
Comment (rjchow) [inline packages/js/remote-logging/src/test/utilts.ts]: filename seems a little weird 🫣

File: packages/js/remote-logging/REDAME.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+- Remote error logging with stack trace analysis
+- Customizable log severity levels
+- Rate limiting to prevent API flooding
+- Automatic capture of unhandled errors and promise rejections
+- Filtering of errors based on WooCommerce asset paths
+- Extensibility through WordPress filters
+
+## Installation
+
+Install the package using npm:
```
Comment (samueljseay) [inline packages/js/remote-logging/REDAME.md:18]: This package is not installable via npm, we should not include this information

**Review by rjchow — APPROVED** [review]
Left a suggestion, otherwise looks good and tests well! Nice work!

File: packages/js/remote-logging/src/test/utilts.ts (hunk 3 above)
Comment (chihsuan) [reply packages/js/remote-logging/src/test/utilts.ts]: Good catch!!

File: packages/js/remote-logging/src/remote-logger.ts (hunk 1 above)
Comment (chihsuan) [reply packages/js/remote-logging/src/remote-logger.ts]: Yeah, that's a good idea. I'll add a filter for it.

File: packages/js/remote-logging/src/remote-logger.ts (hunk 2 above)
Comment (chihsuan) [reply packages/js/remote-logging/src/remote-logger.ts]: Added a filter in e71849344c0df6a0a6b0b659438e718e25ba2503

File: packages/js/remote-logging/src/test/utilts.ts (hunk 3 above)
Comment (chihsuan) [reply packages/js/remote-logging/src/test/utilts.ts]: Fixed in b5fafd484c38afdf960509a6769a98e36b4e3ea4.

**Review by moon0326 — APPROVED** [review]
Really nice work on this! LGTM and tested well 👍 🚀 

Unrelated question: Do we already have rate-limiting in place on WPCOM?

**Comment by chihsuan** [comment]
> Unrelated question: Do we already have rate-limiting in place on WPCOM?

 Yes, we do.

**Pull request merged**